File: 204.count-primes.py
#
# @lc app=leetcode id=204 lang=python3
#
# [204] Count Primes
#

# @lc code=start
class Solution:
    def countPrimes(self, n: int) -> int:
        if n < 2:
            return 0
        nums = [1] * n
        nums[0] = 0
        nums[1] = 0

        def set_flag(i):
            index = i * 2
            while index < n:
                nums[index] = 0
                index += i

        for i in range(2,n):
            if nums[i]:
                set_flag(i)
        # print(nums)
        return len([num for num in nums if num])
                
        
# @lc code=end

if __name__ =="__main__":
    print(Solution().countPrimes(2))

File: golang/204.count-primes.go
/*
 * @lc app=leetcode.cn id=204 lang=golang
 *
 * [204] Count Primes
 */
package main

import "fmt"

// @lc code=start
var (
	mask []bool
	dp   []int
)

func countPrimes(n int) int {
	return dp[n]

}

func init() {
	count := 5*1000000 + 1
	mask = make([]bool, count)
	dp = make([]int, count)
	dp[0] = 0
	mask[0], mask[1] = true, true
	i := 2
	for i < count {
		for i < count && mask[i] {
			i++
		}
		j := i + i
		for j < count {
			mask[j] = true
			j += i
		}
		i++
	}
	for i := 1; i < count; i++ {
		if !mask[i-1] {
			dp[i] = dp[i-1] + 1
		} else {
			dp[i] = dp[i-1]
		}
	}
}

// @lc code=end

func main() {
	fmt.Println(mask[0:10])
}
